controllers/metal3.io/hostfirmwaresettings_controller.go

```diff
@@ -0,0 +1,448 @@
+		// provisioned, its not necessary on every state change as data from Ironic will be the same.
+		if (newState == metal3v1alpha1.StateReady || newState == metal3v1alpha1.StateProvisioned) && oldState != newState {
+			r.Log.Info("baremetalhost event state update", "oldstate", oldState, "newstate", newState)
+			return true
+		}
+
+		return false
+	}
+
+	_, oldHFS := e.ObjectOld.(*metal3v1alpha1.HostFirmwareSettings)
+	_, newHFS := e.ObjectNew.(*metal3v1alpha1.HostFirmwareSettings)
+	if !(oldHFS || newHFS) {
+		// Don't process if not a hostFirmwareSetting
+		return false
+	}
+
+	// This is a hostfirmwaresettings resource, if the update increased the resource generation then process it
+	// changes to LastUpdated will not increase the resource generation
+	if e.ObjectNew.GetGeneration() != e.ObjectOld.GetGeneration() {
+		r.Log.Info("hostfirmwaresettings resource generation changed, updating")
+		return true
+	}
+
+	return false
+}
+
+// Delete EventHandler
+func (r *HostFirmwareSettingsReconciler) deleteEventHandler(e event.DeleteEvent) bool {
+	r.Log.Info("hostfirmwaresettings in delete handler")
+
+	// When the baremetalhost is deleted the corresponding hostfirmwaresettings resources should also be deleted
+	host, hostExists := e.Object.(*metal3v1alpha1.BareMetalHost)
+	if hostExists {
+		hfs := &metal3v1alpha1.HostFirmwareSettings{}
+		if err := r.Get(context.TODO(), client.ObjectKey{Namespace: host.ObjectMeta.Namespace,
+			Name: host.ObjectMeta.Name}, hfs); err != nil {
+
+			if err = r.Delete(context.TODO(), hfs); err != nil {
+				r.Log.Info("hostfirmwaresettings deletion failed")
+			}
+		}
+	}
+
+	// When the hostfirmwaresettings resource is deleted don't let the reconciler run
+	return false
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *HostFirmwareSettingsReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&metal3v1alpha1.HostFirmwareSettings{}).
+		WithEventFilter(
+			predicate.Funcs{
+				UpdateFunc: r.updateEventHandler,
+				DeleteFunc: r.deleteEventHandler,
+			}).
+		Watches(&source.Kind{Type: &metal3v1alpha1.BareMetalHost{}}, &handler.EnqueueRequestForObject{}, builder.Predicates{}).
+		Complete(r)
+}
+
+// Validate the HostFirmwareSetting Spec against the schema
+func (r *HostFirmwareSettingsReconciler) validateHostFirmwareSettings(info *rInfo, schema *metal3v1alpha1.FirmwareSchema) error {
+
+	for name, val := range info.hfs.Spec.Settings {
+		// Prohibit any Spec settings with "Password"
+		if strings.Contains(name, "Password") {
+			return fmt.Errorf("Cannot set Password field")
+		}
+
+		// The setting must be in the Status
+		if _, ok := info.hfs.Status.Settings[name]; !ok {
+			return fmt.Errorf("Setting %s is not in the Status field", name)
+		}
+
+		// check validity of updated value
+		if schema != nil {
+			if err := schema.CheckSettingIsValid(name, val, schema.Spec.Schema); err != nil {
+				return err
```

Comment:
An idea for a follow-up: collect all errors and report them at once.

Reply:
Yeah that's a good idea to handle multiple invalid settings. I've updated it to handle multiple errors and publish events for each one